### ISSUE TITLE
embind: How to export C++ class in linked library to JS

### ISSUE THREAD

**Issue by kpeeters** [emscripten-core/emscripten#11046]
I have a library `A` which I build to a `A.bc` file, and an application `B` which I compile to `B.bc` and then link to that `A.bc` to produce my final `B.wasm`. All fine. Now I would like to use a C++ class which sits inside `A` from javascript (let's call this `Q`). However, when I do `q = new Module.Q()` I get an error that `Module.Q is not a constructor`.

So even though my `B` code can access `Q` which lives in `A`, I cannot access `Q` directly from javascript.

This used to work in the days of 1.38.28 and earlier, when I would tell emscripten to build a shared library for `A.bc` and passed the `-Wl,--export-all-symbols` to the compiler. Now, with 1.39.13, emscripten enforces all libraries to be static, and the above trick does not work anymore.

So in short: how do I ensure that all symbols in the library `A` remain visible in javascript despite the fact that `A` is linked to `B` and I am actually running `B.wasm`? I have only been able to find how to export individual C functions, not entire C++ classes (and I would prefer to have a single catch-all method to avoid having to spell out all things I want to export). Sorry for wasting bandwidth if it's spelled out somewhere already.

**Comment by sbc100** [comment]
I'm not sure what you mean by  exporting a "export a C++ class"... are you using embind perhaps?  

In general emscripten will only keep alive symbols that are marked as `EMSCRIPTEN_KEEPALIVE` or are included in the `-s EXPORTED_FUNCTIONS` list.   As far as I know this has always been true.

**Comment by kpeeters** [comment]
What I mean is that if I have a library `A` and link that to application `B`, then I still want to be able to access the C++ classes in `A` from the JS side. I can obviously access the classes which are directly declared in `B`, but I cannot access those that come from `A`.

This used to work in the days of 1.38.28, without any `EMSCRIPTEN_KEEPALIVE` or `EXPORTED_FUNCTIONS` (I have living proof of that on my disk here). It was a matter of compiling `A` as a *shared* library, and using `--export-all-symbols` in the linking stage that produced that 'shared library'.

**Comment by kpeeters** [comment]
And yes, sorry, this is with embind (the `--bind` parameter passed both in building `A` and `B`).

**Comment by sbc100** [comment]
The `--export-all-symbols` doesn't make much sense to me since that has never been a supported linker flag with wasm-ld.     That flag is supported by lld but only for ELF and COFF targets.

So I can only assume it was build A as a "shared library" that did it.  Can you show me how you built A as a shared library in order to achieve this?    Did if include the SIDE_MODULE setting?

When you say "I still want to be able to access the C++ classes in A from the JS side" do you mean you access directly the mangled C++ names from JS?   C++ symbols, as far as I know, are only available in JS in their mangled form.

**Comment by sbc100** [comment]
Perhaps a simple example that works with 1.38.28 but not anymore might help here?

**Comment by raj96** [comment]
I think this is what you are looking for
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#classes

@alexnoz has made an example as well following the same instructions as above
https://github.com/alexnoz/wasm-class-sample

**Comment by kpeeters** [comment]
@raj96, no those links only describe single libraries, not the situation I described in my original post, in which there is a library `A` linked to library `B` to produce `B.wasm`, and javascript code which wants to use `Q` which is defined in `A`, not `B`.

What I ended up doing was re-exposing `Q` in the `EMSCRIPTEN_BINDINGS` block of library `B` (even though it has already been exposed in library `A`). Not ideal, but it works.

**Comment by raj96** [comment]
So I assume all `A`, `B` and `Q` are classes, where `B` is dependent on `A` and `A` is dependent on `Q`. Am I correct?

**Comment by kpeeters** [comment]
No, `A` and `B` are libraries, library `A` contains (and exports) class `Q`, which is used by class `B`. The javascript wants to use `Q` too. It is not visible if you simply link `A` to `B` to produce `B.wasm`.

**Comment by sbc100** [comment]
If the libraries in questions are ar archive (`.a` files) then you can force all the objects in the archives to be included using `-Wl,-whole-archive` before and `-Wl,-no-whole-archive` after them on the command line.

**Comment by sbc100** [comment]
Embind stuff that is declared in `EMSCRIPTEN_BINDING` does not show up in exports .   Instead they are create dynamic by a static constructor call (which in turn is called by `_initialize`).   Have you tried accessing your `EMSCRIPTEN_BINDING` from JS once the library is loaded.

**Comment by spirobel** [comment]
yes. I have since switched to compiling to wasi as a target directly without emscripten because I want to have my functions in the imports. 
Anyway thanks a lot for your answer 🙂👍

**Comment by gaimanjing** [comment]
> If the libraries in questions are ar archive (`.a` files) then you can force all the objects in the archives to be included using `-Wl,-whole-archive` before and `-Wl,-no-whole-archive` after them on the command line.

https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#binding-libraries
